share/eds/store.go

```diff
@@ -343,25 +371,31 @@ func (s *Store) Blockstore() bstore.Blockstore {
 func (s *Store) CARBlockstore(
 	ctx context.Context,
 	root share.DataHash,
-) (dagstore.ReadBlockstore, error) {
+) (*BlockstoreCloser, error) {
 	ctx, span := tracer.Start(ctx, "store/car-blockstore")
 	tnow := time.Now()
-	r, err := s.carBlockstore(ctx, root)
+	cbs, err := s.carBlockstore(ctx, root)
 	s.metrics.observeCARBlockstore(ctx, time.Since(tnow), err != nil)
 	utils.SetStatusAndEnd(span, err)
-	return r, err
+	return cbs, err
 }
 
 func (s *Store) carBlockstore(
 	ctx context.Context,
 	root share.DataHash,
-) (dagstore.ReadBlockstore, error) {
+) (*BlockstoreCloser, error) {
 	key := shard.KeyFromString(root.String())
-	accessor, err := s.getCachedAccessor(ctx, key)
+	accessor, err := s.cache.Get(key)
+	if err == nil {
+		return blockstoreCloser(accessor)
+	}
+
+	// if the accessor is not found in the cache, create a new one from dagstore
```

Comment:
If the accessor is not found in the cache, this would mean that some other routine has used it and evicted it. In this case, I think it would be insightful to have a metric for how many accessors we get from cache (_and were used_) and how many new ones were created and not retrieved from cache.

I am having a suspicion that maybe more are newly created than retrieved from cache, in which case, caching (_as is here_) would have minimal performance improvement (_cuz once used, the accessor is evicted_), and would make your second upcoming feature (_`but it could be changed to smarter release based on current readers count. Will be implemented as separate issue.`_) crucial for cache to be effective, but that's if my understanding holds true (_I am not sure when `cache.Remove` is called and by whom_), lmk.

Reply:
Accessor could be not in cache, because of 2 reasons:
- it was never put in cache since the startup
- it was evicted, because the cache was full and another item has replaced this one.

Thank you for metrics suggestion. Those you mentioned are crucial to monitor cache performance and implemented in this PR.
 - cache Get hit/miss. Based on miss rate, we can know how many times new accessor will be loaded (either in cache or on stack for single access).
 - Amount loaded into cache is same as eviction rate (because cache is lru). There is eviction rate metric for that

One of the goals of this rework is to prevent caching everything. Given that cache is very limited in its size, we want to prioritise caching items, that has higher chance to be reused. So others accessors with lower probability of being requested should be not cached, because they could take place of accessors with higher demand. It is achieved by  some accessor are never going through cache, like the once you are reffering to